lib/discovery.js

```diff
@@ -0,0 +1,198 @@
+ * @property {(connection, info) => void} peer
+ * @property {(error: Error) => void} error
+ */
+
+/**
+ * @extends {TypedEmitter<DiscoveryEvents>}
+ */
+export class Discovery extends TypedEmitter {
+  #peers = new Map();
+  #discover = new Map();
+  #dht;
+  #mdns;
+
+  /**
+   * @param {object} options
+   * @param {number} options.port
+   * @param {object} options.keyPair
+   * @param {Buffer} options.keyPair.publicKey
+   * @param {Buffer} options.keyPair.secretKey
+   * @param {boolean|object} [options.dht=true]
+   * @param {boolean|object} [options.mdns=true]
+   */
+  constructor({ keyPair, port, dht = true, mdns = true }) {
+    super();
+
+    if (dht !== false) {
+      this.dhtOptions = typeof dht === "object" ? dht : {};
+      this.dhtOptions.keyPair = keyPair;
+
+      this.#dht = new Hyperswarm(this.dhtOptions);
+      this.#dht.on("connection", this._onDhtPeer.bind(this));
+    }
+
+    if (mdns) {
+      const mdnsOptions = typeof mdns === "object" ? mdns : {};
+      mdnsOptions.keyPair = keyPair;
+      mdnsOptions.port = port;
+      mdnsOptions.name = "mapeo";
+
+      this.#mdns = new LocalPeers(mdnsOptions);
+      this.#mdns.on("connection", this._onMdnsPeer.bind(this));
+    }
+  }
+
+  get peers() {
+    return Array.from(this.#peers.values());
+  }
+
+  _onDhtPeer(connection, info) {
+    const publicKey = connection.remotePublicKey.toString("hex");
+    const peer = this.#peers.get(publicKey);
+
+    if (!peer) {
+      const peerInfo = new PeerInfo({
+        topics: info.topics,
+        host: connection.remoteHost,
+        port: connection.remotePort,
+        discoveryType: "dht",
+        identityPublicKey: publicKey,
+      });
+
+      this.#peers.set(publicKey, {
+        connection,
+        peerInfo,
+      });
+
+      this.emit("peer", connection, peerInfo);
+    } else {
+      this.#dht.leavePeer(publicKey);
+    }
+  }
+
+  _onMdnsPeer(connection, info) {
+    const peer = this.#peers.get(info.identityPublicKey);
+
+    if (!peer) {
+      const peerInfo = new PeerInfo({
+        topics: [info.topic],
+        ...info,
+      });
+
+      this.#peers.set(info.identityPublicKey, {
+        connection,
+        peerInfo,
+      });
+
+      this.emit("peer", connection, peerInfo);
+    } else {
+      peer.addTopic(info.topic);
+    }
+  }
+
+  async join(topic) {
+    const topicString = topic.toString("hex");
+    const discover = {};
+
+    if (!this.#discover.has(topicString)) {
+      discover.topic = topicString;
+
+      if (this.#mdns) {
+        discover.mdns = this.#mdns.join(topic);
+      }
+
+      if (this.#dht) {
+        discover.dht = this.#dht.join(topic, {
+          server: this.dhtOptions.server,
+          client: this.dhtOptions.client,
+        });
+
+        if (this.dhtOptions.server) {
+          await discover.dht.flushed();
+        } else {
+          await this.#dht.flush();
+        }
+      }
+
+      this.#discover.set(topicString, discover);
```

Comment:
I think this needs to be above the two `await`s above, to avoid errors when calling join() on the same topic more than once in quick succession.